pymedphys/_dicom/anonymise/__init__.py

```diff
@@ -64,7 +65,47 @@ def create_filename_from_dataset(ds, dirpath=""):
     return pjoin(dirpath, "{}.{}.dcm".format(mode_prefix, ds.SOPInstanceUID))
 
 
-def anonymise_dataset(  # pylint: disable = inconsistent-return-statements
+def anonymise_dataset(ds, copy_dataset=True, replacement_strategy=None):
+    r"""A simple tool to anonymise a DICOM dataset.
+
+    You can find the list of DICOM keywords that are included in default
+    anonymisation `here <./identifying_keywords.json>`__.
+    These were drawn from `DICOM Supp 142
+    <https://www.dicomstandard.org/supplements/>`__
+
+    **We do not yet claim conformance to any DICOM Application Level
+    Confidentiality Profile**, but plan to be in a position to do so in the
+    not-to-distant future.
+
+    Parameters
+    ----------
+    ds : ``pydicom.dataset.Dataset``
+        The DICOM dataset to be anonymised.
+
+    copy : ``bool``, optional
+        If ``True``, then a copy of ``ds`` is returned.
+
+    replacement_strategy: ``immutables.Map`` (keys are VR, value is dispatch function), optional
+        If left as the default value of ``None``, the hardcode replacement strategy is used.
+        Also contains delete_private_tags, delete_unknown_tags, and identifying_keywords as keys
+
+    Returns
+    -------
+    ds_anon : ``pydicom.dataset.Dataset``
+        An anonymised version of the input DICOM dataset.
+    """
+
+    if replacement_strategy is None:
+        replacement_strategy = get_copy_of_strategy()
+
+    with replacement_strategy.mutate() as strategy_copy:
+        strategy_copy["copy_dataset"] = copy_dataset
+    replacement_strategy = strategy_copy.finish()
+
+    return _anonymise_dataset(ds, replacement_strategy=replacement_strategy)
+
+
+def _anonymise_dataset(
     ds,
     replace_values=True,
     keywords_to_leave_unchanged=(),
```

Comment:
Even if it is private, I think this here should default to None (as with every other parameter that can be overwritten). That way if `keywords_to_leave_unchanged` is not None and `keywords_to_leave_unchanged` is defined within the replacement strategy an error can be raised, as no matter what the user of the function provides to `keywords_to_leave_unchanged` it will be silently overwritten by the contents of `replacement_strategy`.

Reply:
OK.

Reply:
What if this function is adjusted to have its signature be:

```python
def _anonymise_dataset(
    ds,
    replace_values=True,
    keywords_to_leave_unchanged=(),
    delete_private_tags=True,
    delete_unknown_tags=None,
    copy_dataset=True,
):
```

And then it is called within `anonymise_dataset` as:

```python
_anonymise_dataset(
    ds, replace_values=strategy["replace_values"],
    keywords_to_leave_unchanged=strategy["keywords_to_leave_unchanged"],
    delete_private_tags=strategy["delete_private_tags"],
    delete_unknown_tags=strategy["delete_unknown_tags"],
    copy_dataset=copy
)
```

That way the following lines won't be needed:

```python
    with replacement_strategy.mutate() as strategy_copy:
        strategy_copy["copy_dataset"] = copy_dataset
    replacement_strategy = strategy_copy.finish()
```

And none of the following logic would be needed:

```python
    if replacement_strategy is not None:
        if "replace_values" in replacement_strategy:
            replace_values = replacement_strategy["replace_values"]
        if "keywords_to_leave_unchanged" in replacement_strategy:
            keywords_to_leave_unchanged = replacement_strategy[
                "keywords_to_leave_unchanged"
            ]
        if "delete_private_tags" in replacement_strategy:
            delete_private_tags = replacement_strategy["delete_private_tags"]
        if "delete_unknown_tags" in replacement_strategy:
            delete_unknown_tags = replacement_strategy["delete_unknown_tags"]
        if "copy_dataset" in replacement_strategy:
            copy_dataset = replacement_strategy["copy_dataset"]

        if (
            identifying_keywords is None
            and "identifying_keywords" in replacement_strategy
        ):

            identifying_keywords = replacement_strategy["identifying_keywords"]
```

And tests that use the old API can just point to `_anonymise_dataset`, new API can point to `anonymise_dataset`.

What are your thoughts?

Reply:
Scrap that, let's try again, downside of trying to quickly parse a lot of code, easy to make silly mistakes.

What if, for now, this function is left as it is

```python
def _anonymise_dataset(
    ds,
    replace_values=True,
    keywords_to_leave_unchanged=(),
    delete_private_tags=True,
    delete_unknown_tags=None,
    copy_dataset=True,
    replacement_strategy=None,
    identifying_keywords=None,
):
```

And then it is called within `anonymise_dataset` as:

```python
_anonymise_dataset(
    ds, replace_values=strategy["replace_values"],
    keywords_to_leave_unchanged=strategy["keywords_to_leave_unchanged"],
    delete_private_tags=strategy["delete_private_tags"],
    delete_unknown_tags=strategy["delete_unknown_tags"],
    copy_dataset=copy,
	replacement_strategy=strategy,
    identifying_keywords=replacement_strategy["identifying_keywords"],
)
```

I believe this would address my primary issue with this PR and simplify the logic.

What are your thoughts?